File: texture-share-gl-client/src/opengl/gl_shared_image.rs
use super::glad;
use std::{os::fd::OwnedFd, ptr};

use texture_share_ipc::platform::{img_data::ImgFormat, ShmemDataInternal};
#[cfg(target_os = "linux")]
type GlMemoryHandle = OwnedFd;

pub struct GlSharedImage {
	// FBO to render image from/to
	fbo: glad::GLuint,

	// Image memory
	mem: glad::GLuint,

	// Image texture
	texture: glad::GLuint,

	data: GlSharedImageData,
}

pub struct GlSharedImageData {
	pub id: u32,
	pub width: u32,
	pub height: u32,
	pub format: glad::GLenum,
	pub allocation_size: u64,
}

#[repr(C)]
pub struct GlImageExtent {
	pub top_left: [glad::GLint; 2],
	pub bottom_right: [glad::GLint; 2],
}

impl Drop for GlSharedImage {
	fn drop(&mut self) {
		if self.texture != 0 {
			unsafe { glad::glad_glDeleteTextures.unwrap()(1, &self.texture) };
			self.texture = 0;
		}

		if self.mem != 0 {
			unsafe { glad::glad_glDeleteMemoryObjectsEXT.unwrap()(1, &self.mem) };
			self.mem = 0;
		}

		if self.fbo != 0 {
			unsafe { glad::glad_glDeleteFramebuffers.unwrap()(1, &self.fbo) };
			self.fbo = 0;
		}
	}
}

impl GlSharedImageData {
	pub fn from_shmem_img_data(data: &ShmemDataInternal) -> GlSharedImageData {
		GlSharedImageData {
			id: data.handle_id,
			width: data.width,
			height: data.height,
			format: GlSharedImage::get_gl_format(data.format),
			allocation_size: data.allocation_size,
		}
	}
}

unsafe fn check_gl<F: FnOnce() -> T, T>(fcn: F) -> Result<T, glad::GLuint> {
	let ret = fcn();
	match glad::glad_glGetError.unwrap_unchecked()() {
		0 => Ok(ret),
		e => Err(e),
	}
}

impl GlSharedImage {
	pub fn init_gl() -> Result<i32, ()> {
		match unsafe { glad::gladLoadGL() } {
			0 => Err(()),
			v => Ok(v),
		}
	}

	pub fn get_gl_format(img_format: ImgFormat) -> glad::GLenum {
		match img_format {
			ImgFormat::B8G8R8 => glad::GL_BGR,
			ImgFormat::B8G8R8A8 => glad::GL_BGRA,
			ImgFormat::R8G8B8 => glad::GL_RGB,
			ImgFormat::R8G8B8A8 => glad::GL_RGBA,
			ImgFormat::Undefined => glad::GL_NONE,
		}
	}

	pub fn get_gl_internal_format(img_format: ImgFormat) -> glad::GLint {
		match img_format {
			ImgFormat::B8G8R8 => glad::GL_RGB8,
			ImgFormat::B8G8R8A8 => glad::GL_RGBA8,
			ImgFormat::R8G8B8 => glad::GL_RGB8,
			ImgFormat::R8G8B8A8 => glad::GL_RGBA8,
			ImgFormat::Undefined => glad::GL_NONE,
		}
		.try_into()
		.unwrap()
	}

	pub fn get_img_format(gl_format: glad::GLenum) -> ImgFormat {
		match gl_format {
			glad::GL_BGR => ImgFormat::B8G8R8,
			glad::GL_BGRA => ImgFormat::B8G8R8A8,
			glad::GL_RGB => ImgFormat::R8G8B8,
			glad::GL_RGBA => ImgFormat::R8G8B8A8,
			_ => ImgFormat::Undefined,
		}
	}

	pub fn new(
		width: glad::GLsizei,
		height: glad::GLsizei,
		allocation_size: glad::GLuint64,
		format: glad::GLenum,
		internal_format: glad::GLint,
		id: u32,
	) -> Result<GlSharedImage, u32> {
		let mut texture: glad::GLuint = 0;
		unsafe {
			check_gl(|| glad::glad_glGenTextures.unwrap()(1, &mut texture as *mut u32))?;
			check_gl(|| glad::glad_glBindTexture.unwrap()(glad::GL_TEXTURE_2D, texture))?;

			check_gl(|| {
				glad::glad_glTexParameteri.unwrap()(
					glad::GL_TEXTURE_2D,
					glad::GL_TEXTURE_MIN_FILTER,
					glad::GL_NEAREST as i32,
				)
			})?;
			check_gl(|| {
				glad::glad_glTexParameteri.unwrap()(
					glad::GL_TEXTURE_2D,
					glad::GL_TEXTURE_MAG_FILTER,
					glad::GL_NEAREST as i32,
				)
			})?;

			check_gl(|| {
				glad::glad_glTexImage2D.unwrap()(
					glad::GL_TEXTURE_2D,
					0,
					internal_format,
					width,
					height,
					0,
					format,
					glad::GL_UNSIGNED_BYTE,
					ptr::null(),
				)
			})?;
			check_gl(|| glad::glad_glBindTexture.unwrap()(glad::GL_TEXTURE_2D, 0))?;
		};

		let data = GlSharedImageData {
			id,
			width: width as u32,
			height: height as u32,
			format,
			allocation_size,
		};
		Ok(GlSharedImage {
			fbo: 0,
			mem: 0,
			texture,
			data,
		})
	}

	#[cfg(target_os = "linux")]
	pub fn import_handle(
		handle: GlMemoryHandle,
		width: glad::GLsizei,
		height: glad::GLsizei,
		allocation_size: glad::GLuint64,
		format: glad::GLenum,
		internal_format: glad::GLenum,
		id: u32,
	) -> Result<GlSharedImage, glad::GLuint> {
		use std::os::fd::IntoRawFd;

		let mut texture: glad::GLuint = 0;
		let mut memory: glad::GLuint = 0;

		unsafe {
			check_gl(|| glad::glad_glGenTextures.unwrap()(1, &mut texture as *mut _))?;
			check_gl(|| glad::glad_glBindTexture.unwrap()(glad::GL_TEXTURE_2D, texture))?;

			check_gl(|| glad::glad_glCreateMemoryObjectsEXT.unwrap()(1, &mut memory as *mut _))?;
			check_gl(|| {
				glad::glad_glImportMemoryFdEXT.unwrap()(
					memory,
					allocation_size,
					glad::GL_HANDLE_TYPE_OPAQUE_FD_EXT,
					handle.into_raw_fd(),
				)
			})?;
			check_gl(|| {
				glad::glad_glTextureStorageMem2DEXT.unwrap()(
					texture,
					1,
					internal_format,
					width,
					height,
					memory,
					0,
				)
			})?;

			check_gl(|| glad::glad_glBindTexture.unwrap()(glad::GL_TEXTURE_2D, 0))?;
		}

		let data = GlSharedImageData {
			id,
			width: width as u32,
			height: height as u32,
			format,
			allocation_size,
		};
		Ok(GlSharedImage {
			fbo: 0,
			mem: 0,
			texture,
			data,
		})
	}

	pub fn get_data(&self) -> &GlSharedImageData {
		&self.data
	}

	fn blit_image(
		src_texture: glad::GLuint,
		src_target: glad::GLuint,
		src_dimensions: &GlImageExtent,
		dst_texture: glad::GLuint,
		dst_target: glad::GLuint,
		dst_dimensions: &GlImageExtent,
		invert: bool,
		blit_fbo: glad::GLuint,
		prev_fbo: glad::GLuint,
	) -> Result<glad::GLuint, glad::GLuint> {
		let blit_fbo = match blit_fbo {
			0 => unsafe {
				let mut fbo: glad::GLuint = 0;
				check_gl(|| glad::glad_glGenFramebuffers.unwrap()(1, &mut fbo))?;
				fbo
			},
			id => id,
		};

		unsafe {
			// bind the FBO (for both, READ_FRAMEBUFFER_EXT and DRAW_FRAMEBUFFER_EXT)
			check_gl(|| {
				glad::glad_glBindFramebuffer.unwrap_unchecked()(glad::GL_FRAMEBUFFER, blit_fbo)
			})?;

			// Attach the Input texture (the shared texture) to the color buffer in our frame buffer - note texturetarget
			check_gl(|| {
				glad::glad_glFramebufferTexture2D.unwrap_unchecked()(
					glad::GL_READ_FRAMEBUFFER,
					glad::GL_COLOR_ATTACHMENT0,
					src_target,
					src_texture,
					0,
				)
			})?;
			check_gl(|| glad::glad_glReadBuffer.unwrap_unchecked()(glad::GL_COLOR_ATTACHMENT0))?;

			// Attach target texture (the one we write into and return) to second attachment point
			check_gl(|| {
				glad::glad_glFramebufferTexture2D.unwrap_unchecked()(
					glad::GL_DRAW_FRAMEBUFFER,
					glad::GL_COLOR_ATTACHMENT1,
					dst_target,
					dst_texture,
					0,
				)
			})?;

			check_gl(|| glad::glad_glDrawBuffer.unwrap_unchecked()(glad::GL_COLOR_ATTACHMENT1))?;

			// Check read/draw fbo for completeness
			match check_gl(|| {
				glad::glad_glCheckFramebufferStatus.unwrap_unchecked()(glad::GL_FRAMEBUFFER)
			})? {
				glad::GL_FRAMEBUFFER_COMPLETE => {
					if invert {
						// copy one texture buffer to the other while flipping upside down
						check_gl(|| {
							glad::glad_glBlitFramebuffer.unwrap_unchecked()(
								src_dimensions.top_left[0],     // srcX0
								src_dimensions.top_left[1],     // srcY0,
								src_dimensions.bottom_right[0], // srcX1
								src_dimensions.bottom_right[1], // srcY1
								dst_dimensions.top_left[0],     // dstX0
								dst_dimensions.bottom_right[1], // dstY0,
								dst_dimensions.bottom_right[0], // dstX1
								dst_dimensions.top_left[1],     // dstY1
								glad::GL_COLOR_BUFFER_BIT,
								glad::GL_LINEAR,
							)
						})?;
					} else {
						// Do not flip during blit
						check_gl(|| {
							glad::glad_glBlitFramebuffer.unwrap_unchecked()(
								src_dimensions.top_left[0],     // srcX0
								src_dimensions.top_left[1],     // srcY0,
								src_dimensions.bottom_right[0], // srcX1
								src_dimensions.bottom_right[1], // srcY1
								dst_dimensions.top_left[0],     // dstX0
								dst_dimensions.top_left[1],     // dstY0
								dst_dimensions.bottom_right[0], // dstX1
								dst_dimensions.bottom_right[1], // dstY1,
								glad::GL_COLOR_BUFFER_BIT,
								glad::GL_LINEAR,
							)
						})?;
					}
				}
				_ => {}
			}

			// restore the previous fbo - default is 0
			check_gl(|| glad::glad_glDrawBuffer.unwrap_unchecked()(glad::GL_COLOR_ATTACHMENT0))?;
			check_gl(|| {
				glad::glad_glBindFramebuffer.unwrap_unchecked()(glad::GL_FRAMEBUFFER, prev_fbo)
			})?;
		};

		Ok(blit_fbo)
	}

	pub fn recv_blit_image(
		&mut self,
		src_texture: glad::GLuint,
		src_target: glad::GLuint,
		src_dimensions: &GlImageExtent,
		invert: bool,
		prev_fbo: glad::GLuint,
	) -> Result<(), glad::GLuint> {
		let blit_fbo = Self::blit_image(
			src_texture,
			src_target,
			src_dimensions,
			self.texture,
			glad::GL_TEXTURE_2D,
			&GlImageExtent {
				top_left: [0, 0],
				bottom_right: [
					self.data.width as glad::GLint,
					self.data.height as glad::GLint,
				],
			},
			invert,
			self.fbo,
			prev_fbo,
		)?;
		self.fbo = blit_fbo;

		Ok(())
	}

	pub fn send_blit_image(
		&mut self,
		dst_texture: glad::GLuint,
		dst_target: glad::GLuint,
		dst_dimensions: &GlImageExtent,
		invert: bool,
		prev_fbo: glad::GLuint,
	) -> Result<(), glad::GLuint> {
		let blit_fbo = Self::blit_image(
			self.texture,
			glad::GL_TEXTURE_2D,
			&GlImageExtent {
				top_left: [0, 0],
				bottom_right: [
					self.data.width as glad::GLint,
					self.data.height as glad::GLint,
				],
			},
			dst_texture,
			dst_target,
			dst_dimensions,
			invert,
			self.fbo,
			prev_fbo,
		)?;
		self.fbo = blit_fbo;

		Ok(())
	}
}

#[cfg(test)]
mod tests {
	use super::glad;
	use super::GlSharedImage;
	use glfw::fail_on_errors;
	use glfw::Context;

	#[allow(dead_code)]
	struct GlfwData<R> {
		glfw: glfw::Glfw,
		window: glfw::PWindow,
		events: glfw::GlfwReceiver<R>,
	}

	fn _init_gl() -> GlfwData<(f64, glfw::WindowEvent)> {
		let mut glfw = glfw::init(fail_on_errors!()).unwrap();
		glfw.window_hint(glfw::WindowHint::Visible(false));

		// Create a windowed mode window and its OpenGL context
		let (mut window, events) = glfw
			.create_window(300, 300, "Hello this is window", glfw::WindowMode::Windowed)
			.expect("Failed to create GLFW window.");
		window.make_current();

		GlSharedImage::init_gl().unwrap();

		GlfwData {
			glfw,
			window,
			events,
		}
	}

	#[test]
	fn gl_shared_image_init_gl() {
		_init_gl();
	}

	#[test]
	fn gl_shared_image_new() {
		let _gl_context = _init_gl();
		let _gl_shared_image =
			GlSharedImage::new(1, 1, 4, glad::GL_RGBA, glad::GL_RGBA as i32, 0).unwrap();
	}
}

File: texture-share-gl-client/cpp/gl_shared_image/gl_shared_image.cpp
#include "gl_shared_image.h"

#include <GL/gl.h>

#define GL_CHECK(x, e, l)                     \
	{                                         \
		x;                                    \
		if((e = glGetError()) != GL_NO_ERROR) \
			goto l;                           \
	}

GlSharedImage::GlSharedImage()
{
	// glGenFramebuffers(1, &this->_fbo);
}

GlSharedImage::~GlSharedImage()
{
	this->Cleanup();
}

bool GlSharedImage::InitializeGLExternal()
{
	return ExternalHandleGl::LoadGlEXT();
}

// void SharedImageHandleGl::InitializeWithExternal(ExternalHandle::SharedImageInfo &&external_handles)
// {
// 	const GLenum gl_format          = ExternalHandleGl::GetGlFormat(external_handles.format);
// 	const GLenum gl_internal_format = ExternalHandleGl::GetGlInternalFormat(external_handles.format);

// 	return this->InitializeWithExternal(std::move(external_handles.handles), external_handles.width,
// 	                                    external_handles.height, external_handles.handle_id,
// 	                                    external_handles.allocation_size, gl_format, gl_internal_format);
// }

GLenum GlSharedImage::Initialize(GLsizei width, GLsizei height, uint64_t handle_id, GLuint64 allocation_size,
                                 GLenum format, GLenum internal_format)
{
	this->Cleanup();

	GLenum error = GL_NO_ERROR;

	GL_CHECK(glGenTextures(1, &this->_image_texture), error, finish_label);
	GL_CHECK(glBindTexture(SHARED_IMAGE_TEX_TARGET, this->_image_texture), error, finish_label);

	GL_CHECK(glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST), error, finish_label);
	GL_CHECK(glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST), error, finish_label);
	GL_CHECK(glTexImage2D(GL_TEXTURE_2D, 0, internal_format, width, height, 0, format, GL_UNSIGNED_BYTE, nullptr),
	         error, finish_label);

finish_label:
	this->_image_data.Width          = width;
	this->_image_data.Height         = height;
	this->_image_data.Format         = format;
	this->_image_data.AllocationSize = allocation_size;
	this->_image_data.Id             = handle_id;

	glBindTexture(SHARED_IMAGE_TEX_TARGET, 0);

	return error;
}

GLenum GlSharedImage::InitializeWithExternal(ExternalHandle::ShareHandles &&share_handles, GLsizei width,
                                             GLsizei height, uint64_t handle_id, GLuint64 allocation_size,
                                             GLenum format, GLenum internal_format)
{
	this->Cleanup();

	GLenum error = GL_NO_ERROR;

	// Create the texture for the FBO color attachment.
	// This only reserves the ID, it doesn't allocate memory
	GL_CHECK(glGenTextures(1, &this->_image_texture), error, finish_label);
	GL_CHECK(glBindTexture(SHARED_IMAGE_TEX_TARGET, this->_image_texture), error, finish_label);

	// Create the GL identifiers

	// semaphores
	// GL_CHECK(ExternalHandleGl::GenSemaphoresEXT(1, &this->_semaphore_read), error, finish_label);
	// GL_CHECK(ExternalHandleGl::GenSemaphoresEXT(1, &this->_semaphore_write), error, finish_label);

	// memory
	GL_CHECK(ExternalHandleGl::CreateMemoryObjectsEXT(1, &this->_mem), error, finish_label);

	// Platform specific import.
	// ExternalHandleGl::ImportSemaphoreExt(this->_semaphore_read, ExternalHandleGl::GL_HANDLE_TYPE,
	//                                      share_handles.ext_read);
	// share_handles.ext_read  = -1;
	// ExternalHandleGl::ImportSemaphoreExt(this->_semaphore_write, ExternalHandleGl::GL_HANDLE_TYPE,
	//                                      share_handles.ext_write);
	// share_handles.ext_write = -1;
	GL_CHECK(ExternalHandleGl::ImportMemoryExt(this->_mem, allocation_size, ExternalHandleGl::GL_HANDLE_TYPE,
	                                           share_handles.memory),
	         error, finish_label);
	share_handles.memory = -1;

	// Use the imported memory as backing for the OpenGL texture.  The internalFormat, dimensions
	// and mip count should match the ones used by Vulkan to create the image and determine it's memory
	// allocation.
	GL_CHECK(ExternalHandleGl::TextureStorageMem2DEXT(this->_image_texture, 1, internal_format, width, height,
	                                                  this->_mem, 0),
	         error, finish_label);

finish_label:
	this->_image_data.Width          = width;
	this->_image_data.Height         = height;
	this->_image_data.Format         = format;
	this->_image_data.Id             = handle_id;
	this->_image_data.AllocationSize = allocation_size;

	glBindTexture(SHARED_IMAGE_TEX_TARGET, 0);

	return error;
}

void GlSharedImage::Cleanup()
{
	if(this->_image_texture > 0)
	{
		glDeleteTextures(1, &this->_image_texture);
		this->_image_texture = 0;
	}

	if(this->_mem > 0)
	{
		ExternalHandleGl::DeleteMemoryObjectsEXT(1, &this->_mem);
		this->_mem = 0;
	}

	// if(this->_semaphore_write)
	// {
	// 	ExternalHandleGl::DeleteSemaphoresEXT(1, &this->_semaphore_write);
	// 	this->_semaphore_write = 0;
	// }

	// if(this->_semaphore_read)
	// {
	// 	ExternalHandleGl::DeleteSemaphoresEXT(1, &this->_semaphore_read);
	// 	this->_semaphore_read = 0;
	// }

	if(this->_fbo > 0)
	{
		glDeleteFramebuffers(1, &this->_fbo);
		this->_fbo = 0;
	}
}

void GlSharedImage::RecvBlitImage(GLuint src_texture_id, GLuint src_texture_target, const ImageExtent &src_dimensions,
                                  bool invert, GLuint prev_fbo)
{
	return this->BlitImage(src_texture_id, src_texture_target, src_dimensions, this->_image_texture,
	                       SHARED_IMAGE_TEX_TARGET,
	                       ImageExtent{
							   {0,                       0                       },
							   {this->_image_data.Width, this->_image_data.Height}
    },
	                       invert, prev_fbo);
}

void GlSharedImage::SendBlitImage(GLuint dst_texture_id, GLuint dst_texture_target, const ImageExtent &dst_dimensions,
                                  bool invert, GLuint prev_fbo)
{
	return this->BlitImage(this->_image_texture, SHARED_IMAGE_TEX_TARGET,
	                       ImageExtent{
							   {0,                       0                       },
							   {this->_image_data.Width, this->_image_data.Height}
    },
	                       dst_texture_id, dst_texture_target, dst_dimensions, invert, prev_fbo);
}

void GlSharedImage::ClearImage(const unsigned char *clear_color)
{
	return this->ClearImage(clear_color, this->_image_data.Format, GL_UNSIGNED_BYTE);
}

void GlSharedImage::ClearImage(const void *clear_color, GLenum format, GLenum type)
{
	glClearTexImage(this->_image_texture, 0, format, type, clear_color);
}

void GlSharedImage::BlitImage(GLuint src_texture_id, GLuint src_texture_target, const ImageExtent &src_dimensions,
                              GLuint dst_texture_id, GLuint dst_texture_target, const ImageExtent &dst_dimensions,
                              bool invert, GLuint prev_fbo)
{
	if(this->_fbo == 0)
	{
		glGenFramebuffers(1, &this->_fbo);
	}

	// bind the FBO (for both, READ_FRAMEBUFFER_EXT and DRAW_FRAMEBUFFER_EXT)
	glBindFramebuffer(GL_FRAMEBUFFER, this->_fbo);

	// Attach the Input texture (the shared texture) to the color buffer in our frame buffer - note texturetarget
	glFramebufferTexture2D(GL_READ_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, src_texture_target, src_texture_id, 0);
	glReadBuffer(GL_COLOR_ATTACHMENT0_EXT);

	// Attach target texture (the one we write into and return) to second attachment point
	glFramebufferTexture2D(GL_DRAW_FRAMEBUFFER, GL_COLOR_ATTACHMENT1, dst_texture_target, dst_texture_id, 0);

	glDrawBuffer(GL_COLOR_ATTACHMENT1);

	// Check read/draw fbo for completeness
	GLuint status = glCheckFramebufferStatus(GL_FRAMEBUFFER);
	if(status == GL_FRAMEBUFFER_COMPLETE_EXT)
	{
		//		if (m_bBLITavailable)
		//		{
		// Flip if the user wants that
		if(!invert)
		{
			// Do not flip during blit
			glBlitFramebuffer(src_dimensions.top_left[0], src_dimensions.top_left[1],         // srcX0, srcY0,
			                  src_dimensions.bottom_right[0], src_dimensions.bottom_right[1], // srcX1, srcY1
			                  dst_dimensions.top_left[0], dst_dimensions.top_left[1],         // dstX0, dstY0,
			                  dst_dimensions.bottom_right[0], dst_dimensions.bottom_right[1], // dstX1, dstY1
			                  GL_COLOR_BUFFER_BIT, GL_LINEAR);
		}
		else
		{
			// copy one texture buffer to the other while flipping upside down
			glBlitFramebuffer(src_dimensions.top_left[0], src_dimensions.top_left[1],         // srcX0, srcY0,
			                  src_dimensions.bottom_right[0], src_dimensions.bottom_right[1], // srcX1, srcY1
			                  dst_dimensions.top_left[0], dst_dimensions.bottom_right[1],     // dstX0, dstY0,
			                  dst_dimensions.bottom_right[0], dst_dimensions.top_left[1],     // dstX1, dstY1
			                  GL_COLOR_BUFFER_BIT, GL_LINEAR);
		}
		//	    }
		//	    else {
		//		    // No fbo blit extension available
		//		    // Copy from the fbo (shared texture attached) to the dest texture
		//		    glBindTexture(TextureTarget, TextureID);
		//			glCopyTexSubImage2D(TextureTarget, 0, 0, 0, 0, 0, width, height);
		//			glBindTexture(TextureTarget, 0);
		//	    }
		//	    bRet = true;
	}
	else
	{
		//		PrintFBOstatus(status);
		//		bRet = false;
	}

	//	std::array<uint8_t, 1920 * 1080 * 4> data;
	//	data.fill(0);
	//	glGetTextureImage(src_texture_id, 0, GL_BGRA, GL_UNSIGNED_BYTE, data.size(), data.data());
	//	code = glGetError();

	//	std::array<uint8_t, 1920 * 1080 * 4> rec_data;
	//	rec_data.fill(0);
	//	glGetTextureImage(src_texture_id, 0, GL_BGRA, GL_UNSIGNED_BYTE, rec_data.size(), rec_data.data());
	//	code = glGetError();

	//	int unequal = 0;
	//	for(size_t i = 0; i < data.size(); ++i)
	//	{
	//			if(data.at(i) != rec_data.at(i))
	//				++unequal;
	//	}

	// restore the previous fbo - default is 0
	glDrawBuffer(GL_COLOR_ATTACHMENT0_EXT); // 04.01.16
	glBindFramebuffer(GL_FRAMEBUFFER, prev_fbo);
}
